Source/Dafny/Verifier/Translator.cs

```diff
@@ -9744,7 +9744,12 @@ public ForceCheckToken(IToken tok)
         Type localType = rhsTypeConstraint;  // this is a type that is appropriate for capturing the value of the RHS
         var ty = TrType(localType);
         var nm = CurrentIdGenerator.FreshId("$rhs#");
-        Bpl.Expr wh = GetWhereClause(tok, new Bpl.IdentifierExpr(tok, nm, ty), localType, etran, NOALLOC);
+        Bpl.Expr wh;
+        if (rhs is HavocRhs && localType.IsNonempty) {
+          wh = GetWhereClause(tok, new Bpl.IdentifierExpr(tok, nm, ty), localType, etran, NOALLOC);
+        } else {
+          wh = null;
```

Comment:
Could you add a comment to the else branch to explain why the where is not needed. If I understand correctly, because the variable is short lived, we know it will not be havoc'ed by Boogie, so a where wouldn't provide additional information over the assigned value.

Btw, are we not possibly missing information for possibly empty types, by leaving out the where? I know putting in the where is unsound, but won't leaving it out be confusing for developers? Would it be an option not to allow havoc'ing a possibly empty type? If doing this havoc leaves you with a variable that is unusable unless you assign to it again, why allow it?

Reply:
Comment added.

I'm looking into one case (that's not likely to ever occur) where the PR so far provides too little information for empty types.

Reply:
I fixed up the remaining case. We are no longer providing too little information for possibly empty types. In particular, when temporary `rhs` variable is used for an assignment of `*`, then that `rhs` variable gets a Boogie `where` clause whose antecedent is the definite-assignment tracker of the variable being assigned. For example, if the Dafny program defines a variable `p` of a possibly empty type, then the assignment

``` dafny
x, p := E, *;
```

is translated into Boogie as:

```
var x: ...;
var defass#p: bool;
var p: P where defass#p ==> W(p); // for some type P and expression W(p)
var rhs0: ...;
var rhs1: P where defass#p ==> W(rhs1); // note, the use of defass#p here
...
rhs0 := E;
havoc rhs1;
x := rhs0;
p := rhs1;
```

Reply:
What does it matter that you've added `where defass#p ==> W(rhs1)` ? That could only effect visible variables after `p := rhs1` right? But given that we also have `where defass#p ==> W(p)`, shouldn't that provide no additional information?

Reply:
If `defass#p` is already `true` at the time of the `havoc rhs1`, then this new `where` clause will have the effect of assuming `W(rhs1)`. This let the verifier prove the last `assert` statement in method `HavocMultipleVar` of `git-issue-1619.dfy`. In my original PR, that assertion did not verify -- which your comment about us possibly "not providing too little information" led me to realize.

Reply:
I see, thanks :)